File: Visual Studio LIghtswitch 2012/Chapter10/HelpDeskVB/HelpDeskVB/Common/UserCode/Engineer.vb
'You can use and redistribute the code from this book (and sample application) for non-commercial and 
'most commercial purposes as long as you acknowledge the source and authorship. 
'You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
'The acknowledgment should include author, title, publisher, and ISBN. 
'An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

Imports System.Text.RegularExpressions

Namespace LightSwitchApplication

    Public Class Engineer

        Private Sub Fullname_Compute(ByRef result As String)
            result = String.Format("{0} - {1}", Surname, Firstname)
        End Sub

        Private Sub Age_Compute(ByRef result As Integer)
            result = DateDiff(DateInterval.Year, DateOfBirth, Now)
        End Sub

        'Listing 5-1. Creating a Validation Warning  
        Private Sub EmailAddress_Validate(results As EntityValidationResultsBuilder)
            ' results.AddPropertyError("<Error-Message>")
            If String.IsNullOrEmpty(EmailAddress) Then
                results.AddPropertyResult(
                   "Providing an Email Address is recommended",
                     ValidationSeverity.Informational)
            End If

        End Sub

        'Listing 5-4. Regex Validation to Check Social Security Numbers
        Private Sub SSN_Validate(results As EntityValidationResultsBuilder)
            ' results.AddPropertyError("<Error-Message>")
            Dim pattern As String =
    "^(?!000)([0-6]\d{2}|7([0-6]\d|7[012]))([ -]?)(?!00)\d\d\3(?!0000)\d{4}$"
            If (Not SSN Is Nothing) AndAlso
               (Regex.IsMatch(SSN, pattern) = False) Then
                results.AddPropertyError(
                   "Enter SSN in format 078-05-1120")
            End If


        End Sub

        'Listing 5-5. Validating File Sizes
        Private Sub EngineerPhoto_Validate(results As EntityValidationResultsBuilder)
            ' results.AddPropertyError("<Error-Message>")
            If Me.EngineerPhoto IsNot Nothing Then
                Dim sizeInKB = Me.EngineerPhoto.Length / 1024
                If sizeInKB > 512 Then
                    results.AddPropertyError("Image Size cannot be > 512kb")
                End If
            End If

        End Sub

        'Listing 5-7. Enforcing Unique Records
        Private Sub ClearanceReference_Validate(results As EntityValidationResultsBuilder)
            ' results.AddPropertyError("<Error-Message>")
            If Len(Me.ClearanceReference) > 0 Then
                Dim duplicateOnServer = (
                    From eng In
                    Me.DataWorkspace.ApplicationData.Engineers.Cast(Of Engineer)()
                    Where
                    eng.Id <> Me.Id AndAlso
                    eng.ClearanceReference.Equals(Me.ClearanceReference,
                        StringComparison.CurrentCultureIgnoreCase)
                    ).ToArray()

                Dim duplicateOnClients = (
                    From eng In
                    Me.DataWorkspace.ApplicationData.Details.GetChanges().
                       OfType(Of Engineer)()
                    Where
                    eng IsNot Me AndAlso
                    eng.ClearanceReference.Equals(Me.ClearanceReference,
                       StringComparison.CurrentCultureIgnoreCase)
                    ).ToArray()

                Dim deletedOnClient = Me.DataWorkspace.ApplicationData.Details.GetChanges().
                   DeletedEntities.OfType(Of Engineer)().ToArray()

                Dim anyDuplicates = duplicateOnServer.Union(duplicateOnClients).Distinct().
                   Except(deletedOnClient).Any()

                If anyDuplicates Then
                    results.AddPropertyError("The clearance reference already exists")
                End If

            End If

        End Sub
    End Class

End Namespace

File: Visual Studio LIghtswitch 2012/Chapter17/HelpDeskCS/HelpDeskCS/Common/UserCode/Engineer.cs
//You can use and redistribute the code from this book (and sample application) for non-commercial and 
//most commercial purposes as long as you acknowledge the source and authorship. 
//You should note the source of the code in any documentation as well as in the program code itself (as a comment). 
//The acknowledgment should include author, title, publisher, and ISBN. 
//An example of such an acknowledgment would be "Derived from Listing 2-10, LightSwitch 2012 by Tim Leung, published by Apress, ISBN 9781430250715".

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Microsoft.LightSwitch;
using System.Text.RegularExpressions;

namespace LightSwitchApplication
{
    public partial class Engineer
    {

        partial void FullName_Compute(ref string result)
        {
            if (Firstname != null && Surname != null)
            {
                result = String.Format("{0} - {1}",
                    Surname.ToString(), Firstname.ToString());
            }
        }

        partial void Age_Compute(ref int result)
        {
            DateTime now = DateTime.Today;
            int age = now.Year - DateOfBirth.Year;
            if (DateOfBirth > now.AddYears(-age)) age--;
            result = age;
        }

        partial void IssueCount_Compute(ref int result)
        {
            result = this.Issues.Count();
        }

        //Listing 5-1. Creating a Validation Warning  
        partial void EmailAddress_Validate(EntityValidationResultsBuilder results)
        {
            // results.AddPropertyError("<Error-Message>");
            if (System.String.IsNullOrEmpty(EmailAddress))
            {
                results.AddPropertyResult(
                   "Providing an Email Address is recommended",
                   ValidationSeverity.Informational);
            }
        }

        //Listing 5-4. Regex Validation to Check Social Security Numbers
        partial void SSN_Validate(EntityValidationResultsBuilder results)
        {
            // results.AddPropertyError("<Error-Message>");
            string pattern =
   @"^(?!000)([0-6]\d{2}|7([0-6]\d|7[012]))([ -]?)(?!00)\d\d\3(?!0000)\d{4}$";
            if (SSN != null && !Regex.IsMatch(SSN, pattern))
            {
                results.AddPropertyError(
                    "Enter SSN in format 078-05-1120");
            }
        }

        //Listing 5-5. Validating File Sizes
        partial void EngineerPhoto_Validate(EntityValidationResultsBuilder results)
        {
            // results.AddPropertyError("<Error-Message>");
            if (EngineerPhoto != null)
            {
                var sizeInKB = EngineerPhoto.Length / 1024;
                if (sizeInKB > 512)
                {
                    results.AddPropertyError("Image Size cannot be > 512kb");
                }
            }

        }

        // Listing 5-7. Enforcing Unique Records
        partial void ClearanceReference_Validate(EntityValidationResultsBuilder results)
        {
            if (ClearanceReference != null &&
                ClearanceReference.Length > 0)
            {
                var duplicatesOnServer = (
                    from eng in
                        this.DataWorkspace.ApplicationData.Engineers.Cast<Engineer>()
                    where (eng.Id != this.Id) &&
                    eng.ClearanceReference.Equals(this.ClearanceReference,
                        StringComparison.CurrentCultureIgnoreCase)
                    select eng
                        ).ToArray();

                var duplicatesOnClient = (
                    from eng in
                        this.DataWorkspace.ApplicationData.Details.GetChanges().
                       OfType<Engineer>()
                    where (eng != this) &&
                    eng.ClearanceReference.Equals(this.ClearanceReference,
                       StringComparison.CurrentCultureIgnoreCase)
                    select eng
                        ).ToArray();

                var deletedOnClient =
                   this.DataWorkspace.ApplicationData.Details.GetChanges().
                      DeletedEntities.OfType<Engineer>().ToArray();

                var anyDuplicates =
                   duplicatesOnServer.Union(duplicatesOnClient).
                      Distinct().Except(deletedOnClient).Any();

                if (anyDuplicates)
                {
                    results.AddPropertyError(
                        "The clearance reference already exists");
                }
            }
        }

        //Listing 17-11. Bypassing Validation
        partial void Engineer_AllowSaveWithErrors(ref bool result)
        {
            result = Application.User.HasPermission(Permissions.CanBypassValidation);
        }

    }
}
